swim/state_transitions.go

```diff
@@ -0,0 +1,198 @@
+
+	// transition contains the transition to execute
+	transition func()
+}
+
+// stateTransitions handles the timers for state transitions in SWIM
+type stateTransitions struct {
+	sync.Mutex
+
+	node *Node
+
+	timeouts StateTimeouts
+	timers   map[string]*transitionTimer
+	enabled  bool
+	logger   log.Logger
+}
+
+// StateTimeouts contains the timeouts for every state before transitioning to the new state
+type StateTimeouts struct {
+	// Suspect is the timeout it takes a node in suspect mode to transition to faulty
+	Suspect time.Duration
+}
+
+// GetTimeout returns the timeout to apply on a transition from state
+func (s *StateTimeouts) GetTimeout(state string) time.Duration {
+	switch state {
+	case Suspect:
+		return s.Suspect
+	default:
+		panic(fmt.Sprintf("unknown state %q", state))
+	}
+}
+
+func mergeStateTimeouts(one StateTimeouts, two StateTimeouts) StateTimeouts {
+	return StateTimeouts{
+		Suspect: util.SelectDuration(one.Suspect, two.Suspect),
+	}
+}
+
+// newStateTransitions returns a new state transition controller that can be used to schedule state transitions for nodes
+func newStateTransitions(n *Node, timeouts StateTimeouts) *stateTransitions {
+	return &stateTransitions{
+		node:     n,
+		timeouts: timeouts,
+		timers:   make(map[string]*transitionTimer),
+		enabled:  true,
+		logger:   logging.Logger("stateTransitions").WithField("local", n.Address()),
+	}
+}
+
+// ScheduleSuspectToFaulty starts the suspect timer. After the Suspect timeout the node will be declared faulty
+func (s *stateTransitions) ScheduleSuspectToFaulty(subject subject) {
+	s.Lock()
+	s.schedule(subject, Suspect, func() {
+		// transition the subject to faulty
+		s.node.memberlist.MakeFaulty(subject.address(), subject.incarnation())
+	})
+	s.Unlock()
+}
+
+func (s *stateTransitions) schedule(subject subject, state string, transition func()) {
+	if !s.enabled {
+		s.logger.WithField("member", subject.address()).Warn("cannot schedule a state transition while disabled")
+		return
+	}
+
+	if s.node.Address() == subject.address() {
+		s.logger.WithField("member", subject.address()).Warn("cannot schedule a state transition for the local member")
+		return
+	}
+
+	if timer, ok := s.timers[subject.address()]; ok {
+		if timer.state == state {
+			s.logger.WithField("member", subject.address()).Warnf("redundant call to schedule a state transition for memer in %q state, ignored", state)
+			return
+		}
+		// cancel the previously scheduled transition for the subject
+		timer.Stop()
+	}
+
+	timer := time.AfterFunc(s.timeouts.GetTimeout(state), func() {
+		s.logger.WithField("member", subject.address()).Infof("executing scheduled transition for member in %q state", state)
+		// execute the transition
+		transition()
+	})
+
+	s.timers[subject.address()] = &transitionTimer{
+		Timer:      timer,
+		state:      state,
+		transition: transition,
+	}
+
+	s.logger.WithField("member", subject.address()).Debugf("scheduled state transition for member in %q state", state)
```

Comment:
I think we should put the state in a field and have a static message. It would be extra nice to include both the start state and the target state.